src/V3DfgPatternPrinter.h

```diff
@@ -0,0 +1,168 @@
+        if (pair.second) pair.first->second += toLetters(m_internedVars.size() - 1);
+        return pair.first->second;
+    }
+
+    const std::string& internSelLsb(uint32_t value) {
+        const auto pair = m_internedSelLsbs.emplace(value, "idx");
+        if (pair.second) pair.first->second += toLetters(m_internedSelLsbs.size() - 1);
+        return pair.first->second;
+    }
+
+    const std::string& internWidth(uint32_t value, const char* base) {
+        const auto pair = m_internedWidth.emplace(value, base);
+        if (pair.second) pair.first->second += toLetters(m_internedWidth.size() - 1);
+        return pair.first->second;
+    }
+
+    // Render the vertx into ss, and return true if the recursion reached the given depth,
+    // meaning an S-expression with that nesting level has been rendered.
+    bool render(std::ostringstream& ss, const DfgVertex& vtx, uint32_t depth) {
+        bool deep = depth == 0;
+
+        if (const DfgConst* const constp = vtx.cast<DfgConst>()) {
+            // Base case 1: constant
+            if (constp->isZero()) {
+                ss << "'0";
+            } else if (constp->isOnes()) {
+                ss << "'1";
+            } else {
+                ss << internConst(*constp);
+            }
+        } else if (const DfgVertexVar* const varp = vtx.cast<DfgVertexVar>()) {
+            // Base case 2: variable
+            ss << internVar(*varp);
+        } else if (depth == 0) {
+            // Base case 3: deep vertex
+            ss << "_";
+        } else {
+            // Recursively print an S-expression for the vertex
+
+            // S-expression begin
+            ss << '(';
+            // Name
+            ss << vtx.typeName();
+            // Specials
+            if (const DfgSel* const selp = vtx.cast<DfgSel>()) {
+                ss << '@';
+                if (selp->lsb() == 0) {
+                    ss << "idx0";
+                } else {
+                    ss << internSelLsb(selp->lsb());
+                }
+            }
+            // Operands
+            vtx.forEachSource([&](const DfgVertex& src) {
+                ss << " ";
+                if (render(ss, src, depth - 1)) deep = true;
+            });
+            // S-expression end
+            ss << ')';
+
+            // Mark it if it has multiple sinks
+            if (vtx.hasMultipleSinks()) ss << "*";
+        }
+
+        // Annotate width
+        ss << '/';
+        const uint32_t width = vtx.width();
+        if (width == 1) {
+            ss << "1";
+        } else if (width <= VL_BYTESIZE) {
+            ss << internWidth(width, "c");
+        } else if (width <= VL_SHORTSIZE) {
+            ss << internWidth(width, "s");
+        } else if (width <= VL_IDATASIZE) {
+            ss << internWidth(width, "i");
+        } else if (width <= VL_QUADSIZE) {
+            ss << internWidth(width, "q");
+        } else {
+            ss << internWidth(width, "w");
+        }
```

Comment:
Perhaps use AstNodeDType::dumpSmall, or make another version next to AstNodeDType::dumpSmall?

Reply:
I will try how that looks, but I'm a bit worried that would reduce the usefulness. Interning the widths helps showing more generic patterns towards the top `(ADD 2-bits + 0)` and (ADD 3-bits + 0)` is the same pattern, so preferably should be printed as 2 instances of the same, not 1 instances each of 2 different patterns.

All this is a bit arbitrary though, will check if it remains useful with your suggestion.

Reply:
However looks best, main point is I'd like it to live in sources next to the existing function.

Reply:
Using `AstNodeDType::dumpSmall` does double the number of unique patterns. However, what we mostly care about is whether expressions are 1 bit, fit in a machine word, or are wide, so I went ahead and used those 3 categories for which vtx.width() is valid, and `AstNodeDType::dumpSmall` for everything else. This has a good balance of brevity and coalescing similar patterns without hiding information.